vortex-serde/src/file/reader/column.rs

```diff
@@ -0,0 +1,124 @@
+use std::cmp::Ordering;
+use std::collections::VecDeque;
+use std::sync::Arc;
+
+use bytes::{Bytes, BytesMut};
+use vortex::array::chunked::ChunkedArray;
+use vortex::compute::slice;
+use vortex::{Array, IntoArray};
+use vortex_dtype::DType;
+use vortex_error::{VortexError, VortexResult};
+
+use crate::file::layouts::Layout;
+use crate::io::VortexReadAt;
+use crate::{ArrayBufferReader, ReadResult};
+
+pub(super) struct ColumnReader {
+    #[allow(dead_code)]
+    name: Arc<str>,
+    dtype: DType,
+    layouts: VecDeque<Layout>,
+    arrays: VecDeque<Array>,
+}
+
+impl ColumnReader {
+    pub fn new(name: Arc<str>, dtype: DType, layouts: VecDeque<Layout>) -> Self {
+        Self {
+            name,
+            dtype,
+            layouts,
+            arrays: Default::default(),
+        }
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.layouts.is_empty() && self.arrays.is_empty()
+    }
+
+    pub fn buffered_row_count(&self) -> usize {
+        self.arrays.iter().map(|arr| arr.len()).sum()
+    }
+
+    pub async fn load<R: VortexReadAt>(
+        &mut self,
+        reader: &mut R,
+        batch_size: usize,
+        context: Arc<vortex::Context>,
+    ) -> VortexResult<()> {
+        loop {
```

Comment:
you could make this a while loop `while(self.buffered_row_count() < batch_size)`